### PR TITLE
Tree PartII

### PR DESCRIPTION
**Pull request by sunnyfz309**
1. Binary Tree Inorder Traversal(recursion and iteration)
2. Binary Tree preorder Traversal(recursion and iteration)
3. Binary Tree Maximum Path Sum
4. BST Minimum Path Sum(Amazon Interview Question)
5. Tree Amplitude(Amazon Interview Question)

### CODE REVIEW

File: src/binaryTreeMaximumPathSum/BSTMinimumPathSum.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Definition for a binary tree node.
+ * public class TreeNode {
+ *     int val;
+ *     TreeNode left;
+ *     TreeNode right;
+ *     TreeNode(int x) { val = x; }
+ * }
+ * Solve BST Minimum Path Sum by recursion
+ * Time:  O(n), where n is the size of the given tree
```
Comment (ly97) [inline src/binaryTreeMaximumPathSum/BSTMinimumPathSum.java]: Is it a minimum path sum from the root to leaf?

Comment (sunnyfz309) [inline src/binaryTreeMaximumPathSum/BSTMinimumPathSum.java]: yes. I'll update it later

File: src/binaryTreeMaximumPathSum/BSTMinimumPathSum.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * Space: O(logn)
+ */
+public class BSTMinimumPathSum {
+	public static int BSTMinPathSum(TreeNode root) {
+		if (root == null) return 0;
```
Comment (ly97) [inline src/binaryTreeMaximumPathSum/BSTMinimumPathSum.java]: 1. It is better to use if, else if, else, which can help you cover all cases. 
2. Once again, you should consider two aspects for the recursion problem. Don't just copy code, you need to build your own solution in a consistent manner.

termination: null node and leaf node.

if(root == null) return 0;
if(root.left == null && root.right == null) return root.val; // leaf node
else if(root.left == null) // right sub tree is empty
else if(root.right == null) // left sub tree is empty
else // none of subtree is empty

File: src/binaryTreeMaximumPathSum/BinaryTreeMaximumPathSum1.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		return maxSum;
+	}
+	
+	// The return value should be the maximum of the three cases: root.val/root.val+left/root.val+right
+	// consider the case root.val + left + right. 
+    // but this path should never be returned back since each node can be only visited once. 
+    // MaxSum = Math.max( above three cases, root.val+left+right)
+	private static int findMax(TreeNode root) {
+		if(root == null) return 0;
+		int left = Math.max(findMax(root.left), 0);
```
Comment (ly97) [inline src/binaryTreeMaximumPathSum/BinaryTreeMaximumPathSum1.java]: Max(some value, 0) is somehow confusing. See this post: http://www.geeksforgeeks.org/find-maximum-path-sum-two-leaves-binary-tree/ 

Additionally, please keep a record of the reference you used.

File: src/binaryTreeInorderTraversal/BinaryTreeInorderTraversal2.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * Solve Binary Tree Inorder Traversal by iteration and stack
+ * Time: O(n), where n is the size of the given tree
+ * Space: O(n)
+ */
+public class BinaryTreeInorderTraversal2 {
+	public static List<Integer> inorderTraversal(TreeNode root) {
+		List<Integer> res = new ArrayList<Integer>();
+        Stack<TreeNode> stack = new Stack<TreeNode>();
+        
+        while (!stack.empty() || root != null) {
```
Comment (ly97) [inline src/binaryTreeInorderTraversal/BinaryTreeInorderTraversal2.java]: It is weird to use root here. You can use TreeNode node = root then use node. Please my code for both preorder and inorder in an iterative way.

File: src/binaryTreeMaximumPathSum/BinaryTreeMaximumPathSum1.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	public static int maxPathSum(TreeNode root) {
+		dfs(root);
+		return res;
+	}
+	
+	// The return value should be the maximum of the three cases: root.val/root.val+left/root.val+right
+	// consider the case root.val + left + right. 
+    // but this path should never be returned back since each node can be only visited once. 
+    // res = Math.max(above three cases, root.val+left+right)
+	private static int dfs(TreeNode root) {
```
Comment (ly97) [inline src/binaryTreeMaximumPathSum/BinaryTreeMaximumPathSum1.java]: Does the comment match the code?

**Pull request merged**